File: Python/2048.py
# coding=UTF-8

'''
    2048
    ====
        Game 2048.
'''

# Import Python Lib
import numpy as np
import random
import sys
import msvcrt
import os

################################################################################
# Game Class
################################################################################

class Game:

    def __init__(self):

        self.__gameMap = np.zeros((4, 4), dtype = int)
        self.createNewNum()
        self.__stepNum = 0


    def __str__(self):

        strList = [str(j).center(6) if j else ' ' * 6 for i in self.__gameMap for j in i]

        return (
            ' ------ ------ ------ ------ \n'
            '|      |      |      |      |\n'
            '|%s|%s|%s|%s|\n'
            '|      |      |      |      |\n'
            ' ------ ------ ------ ------ \n'
            '|      |      |      |      |\n'
            '|%s|%s|%s|%s|\n'
            '|      |      |      |      |\n'
            ' ------ ------ ------ ------ \n'
            '|      |      |      |      |\n'
            '|%s|%s|%s|%s|\n'
            '|      |      |      |      |\n'
            ' ------ ------ ------ ------ \n'
            '|      |      |      |      |\n'
            '|%s|%s|%s|%s|\n'
            '|      |      |      |      |\n'
            ' ------ ------ ------ ------ \n'
            '\nStep: %d' % (
            strList[0], strList[1], strList[2], strList[3],
            strList[4], strList[5], strList[6], strList[7],
            strList[8], strList[9], strList[10], strList[11],
            strList[12], strList[13], strList[14], strList[15],
            self.__stepNum,
        ))


    def createNewNum(self):

        idxList = [(i, j) for i in range(4) for j in range(4) if self.__gameMap[i, j] == 0]
        x, y = random.choice(idxList)
        self.__gameMap[x, y] = random.choice((1, 2))


    def pressLeft(self):

        moveBool = False

        for numLine in self.__gameMap:

            tarIdx = 0

            for idx in [1, 2, 3]:

                if numLine[idx]:

                    if numLine[tarIdx] == 0:

                        numLine[tarIdx] = numLine[idx]
                        numLine[idx] = 0
                        moveBool = True

                    elif numLine[tarIdx] == numLine[idx]:

                        numLine[tarIdx] *= 2
                        numLine[idx] = 0
                        tarIdx += 1
                        moveBool = True

                    else:

                        tarIdx += 1

                        if idx > tarIdx:
                            numLine[tarIdx] = numLine[idx]
                            numLine[idx] = 0
                            moveBool = True

        self.afterPress(moveBool)


    def pressRight(self):

        moveBool = False

        for numLine in self.__gameMap:

            tarIdx = 3

            for idx in [2, 1, 0]:

                if numLine[idx]:

                    if numLine[tarIdx] == 0:

                        numLine[tarIdx] = numLine[idx]
                        numLine[idx] = 0
                        moveBool = True

                    elif numLine[tarIdx] == numLine[idx]:

                        numLine[tarIdx] *= 2
                        numLine[idx] = 0
                        tarIdx -= 1
                        moveBool = True

                    else:

                        tarIdx -= 1

                        if idx < tarIdx:
                            numLine[tarIdx] = numLine[idx]
                            numLine[idx] = 0
                            moveBool = True

        self.afterPress(moveBool)


    def pressUp(self):

        moveBool = False

        for colIdx in range(4):

            numLine = self.__gameMap[:,colIdx]
            tarIdx = 0

            for idx in [1, 2, 3]:

                if numLine[idx]:

                    if numLine[tarIdx] == 0:

                        numLine[tarIdx] = numLine[idx]
                        numLine[idx] = 0
                        moveBool = True

                    elif numLine[tarIdx] == numLine[idx]:

                        numLine[tarIdx] *= 2
                        numLine[idx] = 0
                        tarIdx += 1
                        moveBool = True

                    else:

                        tarIdx += 1

                        if idx > tarIdx:
                            numLine[tarIdx] = numLine[idx]
                            numLine[idx] = 0
                            moveBool = True

        self.afterPress(moveBool)


    def pressDown(self):

        moveBool = False

        for colIdx in range(4):

            numLine = self.__gameMap[:,colIdx]
            tarIdx = 3

            for idx in [2, 1, 0]:

                if numLine[idx]:

                    if numLine[tarIdx] == 0:

                        numLine[tarIdx] = numLine[idx]
                        numLine[idx] = 0
                        moveBool = True

                    elif numLine[tarIdx] == numLine[idx]:

                        numLine[tarIdx] *= 2
                        numLine[idx] = 0
                        tarIdx -= 1
                        moveBool = True

                    else:

                        tarIdx -= 1

                        if idx < tarIdx:
                            numLine[tarIdx] = numLine[idx]
                            numLine[idx] = 0
                            moveBool = True

        self.afterPress(moveBool)


    def winCheck(self):

        for i in self.__gameMap:
            if 2048 in i:
                return True

        return False


    def loseCheck(self):

        for line in self.__gameMap:

            if 0 in line:
                return False

            for idx in range(3):
                if line[idx] == line[idx + 1]:
                    return False

        for colIdx in range(4):

            colList = self.__gameMap[:, colIdx]

            for idx in range(3):
                if colList[idx] == colList[idx + 1]:
                    return False

        return True


    def afterPress(self, moveBool):

        if moveBool:
            self.__stepNum += 1
            self.createNewNum()

        if self.winCheck():
            os.system('cls')
            print(self)
            print('\nYou Win!')
            os.system('pause')
            sys.exit(0)

        if self.loseCheck():
            os.system('cls')
            print(self)
            print('\nYou Lose!')
            os.system('pause')
            sys.exit(0)


    def Run(self):

        os.system('cls')
        print(self)

        while True:

            msvcrt.getch()
            inputChar = msvcrt.getch()

            if inputChar == b'H':
                self.pressUp()
            elif inputChar == b'P':
                self.pressDown()
            elif inputChar == b'K':
                self.pressLeft()
            elif inputChar == b'M':
                self.pressRight()

            os.system('cls')
            print(self)


################################################################################
# Main Program Define
################################################################################

def main():

    gameObj = Game()
    gameObj.Run()


if __name__ == '__main__':
    main()

File: src/2048.cpp
/*
    Implementation of the 2048 game.
*/

#include "2048.h"

#include <array>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <ctime>
#include <ncurses.h>
#include <vector>

// Namespace Begin
namespace Game2048 {

using namespace std;

////////////////////////////////////////////////////////////////////////////////
// Game Class Init Function
////////////////////////////////////////////////////////////////////////////////

Game::Game() {
  __gameMap = array<array<int, 4>, 4>{};
  __stepNum = 0;
  createNewNum();
}

////////////////////////////////////////////////////////////////////////////////
// Show
////////////////////////////////////////////////////////////////////////////////

void Game::show() const {
  initscr();
  cbreak();
  noecho();
  char strList[16][7];
  int idx = 0;

  for (auto i : __gameMap) {
    for (int j : i) {
      if (j) {
        sprintf(strList[idx], "  %-4d", j);
      } else {
        strcpy(strList[idx], "      ");
      }

      idx++;
    }
  }

  clear();

  printw(" ------ ------ ------ ------ \n");
  printw("|      |      |      |      |\n");
  printw("|%s|%s|%s|%s|\n", strList[0], strList[1], strList[2], strList[3]);
  printw("|      |      |      |      |\n");
  printw(" ------ ------ ------ ------ \n");
  printw("|      |      |      |      |\n");
  printw("|%s|%s|%s|%s|\n", strList[4], strList[5], strList[6], strList[7]);
  printw("|      |      |      |      |\n");
  printw(" ------ ------ ------ ------ \n");
  printw("|      |      |      |      |\n");
  printw("|%s|%s|%s|%s|\n", strList[8], strList[9], strList[10], strList[11]);
  printw("|      |      |      |      |\n");
  printw(" ------ ------ ------ ------ \n");
  printw("|      |      |      |      |\n");
  printw("|%s|%s|%s|%s|\n", strList[12], strList[13], strList[14], strList[15]);
  printw("|      |      |      |      |\n");
  printw(" ------ ------ ------ ------ \n");
  printw("\nStep: %d\n", __stepNum);
}

////////////////////////////////////////////////////////////////////////////////
// Create A New Number
////////////////////////////////////////////////////////////////////////////////

void Game::createNewNum() {
  vector<array<int, 2>> idxList;

  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      if (__gameMap[i][j] == 0) {
        idxList.push_back(array<int, 2>{i, j});
      }
    }
  }

  int listLen = idxList.size();

  srand(time(nullptr));
  int randIdx = rand() % listLen;

  __gameMap[idxList[randIdx][0]][idxList[randIdx][1]] = rand() % 2 + 1;
}

////////////////////////////////////////////////////////////////////////////////
// Press Left
////////////////////////////////////////////////////////////////////////////////

void Game::pressLeft() {
  bool moveBool = false;

  for (auto &numLine : __gameMap) {
    int tarIdx = 0;

    for (int idx = 1; idx < 4; idx++) {
      if (numLine[idx]) {
        if (numLine[tarIdx] == 0) {
          numLine[tarIdx] = numLine[idx];
          numLine[idx] = 0;
          moveBool = true;
        } else if (numLine[tarIdx] == numLine[idx]) {
          numLine[tarIdx] *= 2;
          numLine[idx] = 0;
          tarIdx++;
          moveBool = true;
        } else {
          tarIdx++;

          if (idx > tarIdx) {
            numLine[tarIdx] = numLine[idx];
            numLine[idx] = 0;
            moveBool = true;
          }
        }
      }
    }
  }

  afterPress(moveBool);
}

////////////////////////////////////////////////////////////////////////////////
// Press Right
////////////////////////////////////////////////////////////////////////////////

void Game::pressRight() {
  bool moveBool = false;

  for (auto &numLine : __gameMap) {
    int tarIdx = 3;

    for (int idx = 2; idx >= 0; idx--) {
      if (numLine[idx]) {
        if (numLine[tarIdx] == 0) {
          numLine[tarIdx] = numLine[idx];
          numLine[idx] = 0;
          moveBool = true;
        } else if (numLine[tarIdx] == numLine[idx]) {
          numLine[tarIdx] *= 2;
          numLine[idx] = 0;
          tarIdx--;
          moveBool = true;
        } else {
          tarIdx--;

          if (idx < tarIdx) {
            numLine[tarIdx] = numLine[idx];
            numLine[idx] = 0;
            moveBool = true;
          }
        }
      }
    }
  }

  afterPress(moveBool);
}

////////////////////////////////////////////////////////////////////////////////
// Press Up
////////////////////////////////////////////////////////////////////////////////

void Game::pressUp() {
  bool moveBool = false;

  for (int colIdx = 0; colIdx < 4; colIdx++) {
    array<int *, 4> numPtrList;

    for (int idx = 0; idx < 4; idx++) {
      numPtrList[idx] = &__gameMap[idx][colIdx];
    }

    int tarIdx = 0;

    for (int idx = 1; idx <= 3; idx++) {
      if (*numPtrList[idx]) {
        if (*numPtrList[tarIdx] == 0) {
          *numPtrList[tarIdx] = *numPtrList[idx];
          *numPtrList[idx] = 0;
          moveBool = true;
        } else if (*numPtrList[tarIdx] == *numPtrList[idx]) {
          *numPtrList[tarIdx] *= 2;
          *numPtrList[idx] = 0;
          tarIdx++;
          moveBool = true;
        } else {
          tarIdx++;

          if (idx > tarIdx) {
            *numPtrList[tarIdx] = *numPtrList[idx];
            *numPtrList[idx] = 0;
            moveBool = true;
          }
        }
      }
    }
  }

  afterPress(moveBool);
}

////////////////////////////////////////////////////////////////////////////////
// Press Down
////////////////////////////////////////////////////////////////////////////////

void Game::pressDown() {
  bool moveBool = false;

  for (int colIdx = 0; colIdx < 4; colIdx++) {
    array<int *, 4> numPtrList;

    for (int idx = 0; idx < 4; idx++) {
      numPtrList[idx] = &__gameMap[idx][colIdx];
    }

    int tarIdx = 3;

    for (int idx = 2; idx >= 0; idx--) {
      if (*numPtrList[idx]) {
        if (*numPtrList[tarIdx] == 0) {
          *numPtrList[tarIdx] = *numPtrList[idx];
          *numPtrList[idx] = 0;
          moveBool = true;
        } else if (*numPtrList[tarIdx] == *numPtrList[idx]) {
          *numPtrList[tarIdx] *= 2;
          *numPtrList[idx] = 0;
          tarIdx--;
          moveBool = true;
        } else {
          tarIdx--;

          if (idx < tarIdx) {
            *numPtrList[tarIdx] = *numPtrList[idx];
            *numPtrList[idx] = 0;
            moveBool = true;
          }
        }
      }
    }
  }

  afterPress(moveBool);
}

////////////////////////////////////////////////////////////////////////////////
// Win Check
////////////////////////////////////////////////////////////////////////////////

bool Game::winCheck() const {
  for (auto i : __gameMap) {
    for (int j : i) {
      if (j == 2048) {
        return true;
      }
    }
  }

  return false;
}

////////////////////////////////////////////////////////////////////////////////
// Lose Check
////////////////////////////////////////////////////////////////////////////////

bool Game::loseCheck() const {
  for (auto line : __gameMap) {
    for (int j : line) {
      if (j == 0) {
        return false;
      }
    }

    for (int idx = 0; idx < 3; idx++) {
      if (line[idx] == line[idx + 1]) {
        return false;
      }
    }
  }

  for (int colIdx = 0; colIdx < 4; colIdx++) {
    vector<int> colList;

    for (auto line : __gameMap) {
      colList.push_back(line[colIdx]);
    }

    for (int idx = 0; idx < 3; idx++) {
      if (colList[idx] == colList[idx + 1]) {
        return false;
      }
    }
  }

  return true;
}

////////////////////////////////////////////////////////////////////////////////
// After Press
////////////////////////////////////////////////////////////////////////////////

void Game::afterPress(const bool moveBool) {
  if (moveBool) {
    __stepNum++;
    createNewNum();
  }

  if (winCheck()) {
    show();
    printw("\nYou Win!\n");
    getch();
    exit(0);
  }

  if (loseCheck()) {
    show();
    printw("\nYou Lose!\n");
    getch();
    exit(0);
  }
}

////////////////////////////////////////////////////////////////////////////////
// Run
////////////////////////////////////////////////////////////////////////////////

void Game::Run() {
  initscr();
  cbreak();
  noecho();
  show();
  char inputChar;

  while (1) {
    inputChar = getch();

    if (inputChar == 'H') {
      pressUp();
    } else if (inputChar == 'P') {
      pressDown();
    } else if (inputChar == 'K') {
      pressLeft();
    } else if (inputChar == 'M') {
      pressRight();
    }

    show();
  }
  endwin();
}

// Namespace End
} // namespace Game2048